experimental/tools/cpp-simulation/cpp-simulation.cpp

```diff
@@ -0,0 +1,697 @@
+};
+
+class ChannelState : public TypedValueState<handshake::ChannelType> {
+  // give the corresponding RW API an access to data members
+  friend struct ChannelConsumerRW;
+  friend struct ChannelProducerRW;
+  // give the corresponding Updater an access to data members
+  friend class ChannelUpdater;
+
+public:
+  using TypedValueState<handshake::ChannelType>::TypedValueState;
+
+  ChannelState(TypedValue<handshake::ChannelType> channel)
+      : TypedValueState<handshake::ChannelType>(channel) {
+
+    llvm::TypeSwitch<Type, void>(channel.getType().getDataType())
+        .Case<IntegerType>([&](IntegerType intType) {
+          data = APInt(intType.getWidth(), 0, intType.isSigned());
+        })
+        .Case<FloatType>([&](FloatType floatType) {
+          data = APFloat(floatType.getFloatSemantics());
+        })
+        .Default([&](auto) {
+          emitError(channel.getLoc())
+              << "Unsuported date type " << channel.getType()
+              << ", we111 should probably report an error and stop";
+        });
+  }
+
+protected:
+  bool valid = false;
+  bool ready = false;
+  llvm::Any data = {};
+  // maybe some additional data signals here
+};
+
+class ControlState : public TypedValueState<handshake::ControlType> {
+  // give the corresponding RW API an access to data members
+  friend struct ControlConsumerRW;
+  friend struct ControlProducerRW;
+  // give the corresponding Updater an access to data members
+  friend class ControlUpdater;
+
+public:
+  using TypedValueState<handshake::ControlType>::TypedValueState;
+
+  ControlState(TypedValue<handshake::ControlType> control)
+      : TypedValueState<handshake::ControlType>(control) {}
+
+protected:
+  bool valid = false;
+  bool ready = false;
+};
+
+//===----------------------------------------------------------------------===//
+// Updater
+//===----------------------------------------------------------------------===//
+/// Classes to update old values states with new ones after finishing the data
+/// collection.
+/// The map updaters (defined inside the Simulator class) contains references to
+/// the corresponding pair of new and old valueStates, assigns the newly
+/// collected values from newValueStates to oldValueStates when the collection
+/// is finished.
+
+/// Base Updater class
+class Updater {
+public:
+  virtual ~Updater() = default;
+  virtual void update() = 0;
+};
+
+/// Templated updater class that contains pair <oldValueState, newValueState>
+template <typename State>
+class DoubleUpdater : public Updater {
+protected:
+  State &cons;
+  State &prod;
+
+public:
+  DoubleUpdater(State &cons, State &prod) : cons(cons), prod(prod) {}
+};
+
+/// Update valueStates with "Channel type": valid, ready, data
+class ChannelUpdater : public DoubleUpdater<ChannelState> {
+public:
+  ChannelUpdater(ChannelState &cons, ChannelState &prod)
+      : DoubleUpdater<ChannelState>(cons, prod) {}
+  void update() override {
```

Comment:
I guess this is the function where you will check if an execution model has changed some external state? If so you probably need this function to return something to tell you whether that is the case.